src/internal/ppsutil/util.go

```diff
@@ -363,23 +355,32 @@ func UpdateJobState(pipelines col.ReadWriteCollection, jobs col.ReadWriteCollect
 	}
 	pipelinePtr.JobCounts[int32(state)]++
 	pipelinePtr.LastJobState = state
-	if err := pipelines.Put(jobPtr.Pipeline.Name, pipelinePtr); err != nil {
+	if err := pipelines.Put(jobPtr.Job.Pipeline.Name, pipelinePtr); err != nil {
 		return err
 	}
 
 	// Update job info
 	var err error
 	if state == pps.JobState_JOB_STARTING {
```

Comment:
Where do we actually update the job state to starting? I see where we are setting a timestamp in `startJob` in the registry, which should probably just go through this code path instead. Also, I'm not sure if it makes sense to set the started if it is nil in the terminal case. If we are depending on either both or none being then set, then I guess it is fine.

Reply:
So we have some bits of code which expect that `Started` is set if `Finished` is set (for example, calculating the duration of a job).  I think it's reasonable to ensure that in the data structure.

As for where the job state is set to starting, I think that's actually a bug and it's skipping over the STARTED state in `registry.go` - I'll fix that.